Put the ditto printing code back into the latest Smoothieware.

RTimothyEdwards opened the pull request:
This feature adds "ditto printing" to Smoothieware, which is the ability to print the same object multiple times with a multiple-extruder system having extruder heads on the same X axis belt, separated by a fixed distance.  gcode for a single extruder is sent to the printer, which duplicates the extruder motor commands for all extruders.  The feature can be turned on or off by the M605 command, which in other firmwares serves the same or similar purpose.

Feature description:

Ditto printing works by adding a 1-bit flag to the extruder for "always_on".  This allows an extruder to be put into a state where it cannot be deselected.  It will then duplicate all extruder commands sent to the active extruder.  Command "M605 S2" declares that all extruders that are not the active extruder will follow (duplicate) the actions of the active extruder.  Command "M605 S0" will return to normal operating conditions.

Code changes:

In addition to the extra flag in the extruder record, ditto printing requires that there not be a single "selected extruder" in Robot.cpp, but instead always loop through all extruders and send extruder commands to all that are selected.  Extruder.cpp modifies the deselect() method to prevent an extruder marked "always on" from being deselected.

This code should incur no processing overhead to single-extruder systems, and trivial overhead to dual- (or more) extruder systems.  The additional bit flag fits in the existing extruder record, which was under-utilized for flag bits.

RTimothyEdwards committed:
Put the ditto printing code back into the latest Smoothieware.
Requires not declaring a single "selected extruder" in Robot.cpp,
but always looping through all extruders and sending commands to
all that are enabled.

wolfmanjm commented:
remove all checkins to FirmwareBin as that makes it impossible to merge, this is explained in the contribution guidelines.

RTimothyEdwards committed:
Removed conflicting file from FirmwareBin

RTimothyEdwards committed:
Merge branch 'edge' of https://github.com/Smoothieware/Smoothieware into feature/ditto-printing

Pulling edge

wolfmanjm commented:
Is this really a generally useful feature? It seems to me to be an edge case that could be maintained in a fork of smoothie instead of adding to the core code.
@arthurwolf what do you say. I'd prefer to not accept this.

wolfmanjm commented:
FYI you have to remove all commits to any file in FirmwareBin, this directory is NOT for users to put stuff in.

RTimothyEdwards commented:
I didn't understand that.  I removed the firmware.bin file from FirmwareBin but if that is not sufficient, I don't know how to put the contents back the way they were without redoing everything from the beginning.

RTimothyEdwards commented:
And yes, ditto printing is a generally useful feature.  At any rate, removing the restriction that only one extruder can be active at a given time allows for all sorts of other possibilities, as well, such as speeding up dual extruder printing by retracting one head at the same time the other is being extruded.  Currently the code does not allow for this.  Saving time is critical for dual printing, and most of the firmwares in existence are horribly inefficient for any kind of dual extruder printing.

RTimothyEdwards commented:
Also, M605 is a known documented reprap command (http://reprap.org/wiki/G-code). Although it is principally directed at dual X-carriage movement, the option "M605 S2" is specifically for having a second extruder duplicate the first, i.e., ditto printing.  My implementation is compatible but does not have the optional arguments for distance separation and temperature differential (the first does not apply to a machine with a single X belt, unless the argument is meant to be advisory only; and the second seems unnecessary but would not be difficult to implement if anyone felt like it).

wolfmanjm commented:
You violated the contribution guidelines where it says "Do not commit changes to the FirmwareBin directory or any binary files in the repository." There is a reason for that, and you need to reverse that or it will not be merged.

wolfmanjm commented:
My preference as I stated is that this be maintained in a fork of smoothie. I do not like the intrusive changes made that complicate the code, and make it much harder to maintain. There were also several subtle reasons that only one extruder could be active at a time and I do not think this addresses those issues at the moment.
This will need very careful review and make sure it does not break other aspects of smoothie such as n-axis control.

RTimothyEdwards commented:
(In reference to the accidental changes in FirmwareBin) Yes, I understand that, but I cannot find a way to reverse it.  I tried.  Right now I have those files in my local branch the way they were originally, but it won't let me do a commit. . .

wolfmanjm commented:
That is why binary files are not to be committed :) it is a pain to revert them. I am not sure I know how to fix it.

RTimothyEdwards commented:
It's just a couple of code changes in a couple of files. . . I can do it over again.  But would I need to create a new branch or could I just delete this branch and create it again?

RTimothyEdwards commented:
This is why I don't like git. . .   https://xkcd.com/1597/

wolfmanjm commented:
Yea that is why I added the comment in the contribution guidelines.

wolfmanjm commented:
Here is what I suggest, you add you fork to the third party forks area of the wiki, we can point people at it and they can test it. If enough people can get it to work then it will be reconsidered.

However right off the bat I see issues with various M commands that are targeted at the active extruder, when they are all active they no longer work as expected.
Also I think it breaks G10/G11.
and probably volumetric extrusion.
Just a few things you need to consider.

RTimothyEdwards commented:
Fair enough.
I have added the fork to the 3rd party forks area.
I also deleted the branch and regenerated it (without messing up the FirmwareBin files this time), which has apparently had the effect of automatically closing this thread.

wolfmanjm commented:
Yea that may not be the best way to do it. I am not sure how to recover from that issue otherwise I would have suggested it.

wolfmanjm commented:
As it is there now if anyone asks about that kind of function we can point them to it.

arthurwolf commented:
Yes please add this to the third-party branches on the wiki, talk about it in the various smoothie communities/forums so that users are aware of it, and after a few users have actually used it, we can talk about merging it back. that'll require making sure the change is as small as it can possibly be while still providing the feature.
this is actually generally useful and I'd like it to be merged ultimately, but we need to make sure it doesn't break anything

RTimothyEdwards commented:
Thanks, Arthur.  I'm not exactly sure how to promote it (such as which forum(s) is/are most appropriate) but I will link to it from my own MakerGear twin extruder project page.  I was updating the code yesterday per request from someone who has built their own printer and apparently found my web page when looking for "ditto printing".  So I built the custom version for them.

Of course testing is needed (and I haven't done any testing other than loading the firmware and checking that both extruders run at the same time when the feature is enabled).  The code will not break anything as long as the feature is not enabled.  However, as Jim noted, there are strict caveats when it is enabled.  But then this will always be true, since, for example, you cannot ditto print with a 0.3mm nozzle on the left side and a 0.5mm nozzle on the right,  as the slicing would be completely different for the two sides.  Ditto printing implies that the left and right extruder configurations must be the same.

arthurwolf commented:
I promoted it on Google+'s Smoothie community, we'll see if it gets us a
few users.

On Tue, Feb 21, 2017 at 2:59 PM, R. Timothy Edwards <
notifications@github.com> wrote:

> Thanks, Arthur. I'm not exactly sure how to promote it (such as which
> forum(s) is/are most appropriate) but I will link to it from my own
> MakerGear twin extruder project page. I was updating the code yesterday per
> request from someone who has built their own printer and apparently found
> my web page when looking for "ditto printing". So I built the custom
> version for them.
>
> Of course testing is needed (and I haven't done any testing other than
> loading the firmware and checking that both extruders run at the same time
> when the feature is enabled). The code will not break anything as long as
> the feature is not enabled. However, as Jim noted, there are strict caveats
> when it is enabled. But then this will always be true, since, for example,
> you cannot ditto print with a 0.3mm nozzle on the left side and a 0.5mm
> nozzle on the right, as the slicing would be completely different for the
> two sides. Ditto printing implies that the left and right extruder
> configurations must be the same.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1130#issuecomment-281351591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFailiB7gO2Zek6M-mrj1j8HMCaR0ks5reu3XgaJpZM4MGwX_>
> .
>



-- 
Courage et bonne humeur.
